contracts/src/token/erc20/utils/safe_erc20.rs

```diff
@@ -0,0 +1,225 @@
+    }
+
+    /// Decrease the calling contract's allowance toward `spender` by `requestedDecrease`. If `token` returns no
+    /// value, non-reverting calls are assumed to be successful.
+    pub fn safe_decrease_allowance(
+        &mut self,
+        token: Address,
+        spender: Address,
+        requested_decrease: U256,
+    ) -> Result<(), Error> {
+        let erc20 = IERC20::new(token);
+        let call = Call::new_in(self);
+        let current_allowance =
+            erc20.allowance(call, address(), spender).or({
+                Err(Error::SafeErc20FailedOperation(SafeErc20FailedOperation {
+                    token,
+                }))
+            })?;
+
+        if current_allowance < requested_decrease {
+            return Err(Error::SafeErc20FailedDecreaseAllowance(
+                SafeErc20FailedDecreaseAllowance {
+                    spender,
+                    currentAllowance: current_allowance,
+                    requestedDecrease: requested_decrease,
+                },
+            ));
+        }
+
+        self.force_approve(
+            token,
+            spender,
+            current_allowance - requested_decrease,
+        )
+    }
+
+    /// Set the calling contract's allowance toward `spender` to `value`. If `token` returns no value,
+    /// non-reverting calls are assumed to be successful. Meant to be used with tokens that require the approval
+    /// to be set to zero before setting it to a non-zero value, such as USDT.
+    pub fn force_approve(
+        &mut self,
+        token: Address,
+        spender: Address,
+        value: U256,
+    ) -> Result<(), Error> {
+        let selector = function_selector!("approve", Address, U256);
+
+        // Helper function to construct calldata
+        fn build_approve_calldata(
+            spender: Address,
+            value: U256,
+            selector: &[u8],
+        ) -> Vec<u8> {
+            let encoded_args = (spender, value).abi_encode_params();
+            [&selector[..4], &encoded_args].concat()
+        }
+
+        // Try performing the approval with the desired value
+        let approve_data = build_approve_calldata(spender, value, &selector);
+        if self.call_optional_return(token, &approve_data).is_ok() {
+            return Ok(());
+        }
+
+        // If that fails, reset allowance to zero, then retry the desired approval
+        let reset_data = build_approve_calldata(spender, U256::ZERO, &selector);
+        self.call_optional_return(token, &reset_data)?;
+        self.call_optional_return(token, &approve_data)?;
+
+        Ok(())
+    }
+}
+
+impl SafeErc20 {
+    /// Imitates a Solidity high-level call, relaxing the requirement on the return value:
+    /// if data is returned, it must not be `false`, otherwise calls are assumed to be successful.
+    fn call_optional_return(
+        &self,
+        token: Address,
+        data: &[u8],
+    ) -> Result<(), Error> {
+        match RawCall::new()
```

Comment:
The reason I opted for `RawCall` here is that it enables us to set the return data size to just 32 bytes, which assumes the underlying ERC20 returns a boolean. It also does not make assumptions about the return value of the underlying ERC20.